### DISCUSSION TITLE
Doubt regarding convertion of a quantum circuit into an exportable tensor network

### DISCUSSION THREAD

**Discussion by zadnarock** [jcmgray/quimb#162] [Q&A]
Hey, I am new to all this, can someone explain how to convert a qtn tensor network obtained through a quantum circuit into an adjacency matrix or any other storable form?

**Comment by jcmgray** [comment]
Hi @zadnarock, are you just wanting to write the tensor network to disk (you can just use pickle or e.g. joblib.dump) or use in some other framework?

The most minimal bits of information to be able to reconstruct a tensor network are:
```python
for tensor in tn:
    # incident edges to this node
    inds = tensor.inds
    # the data at each node
    data = tensor.data
```
There is [lots more information here](https://quimb.readthedocs.io/en/latest/tensor-design.html).

The geometry is not always exactly a graph (i.e. has an adjacency matrix), since normal TNs can have output indices (those indices that only appear once). Moreover, if any indices appear on more than two tensors, then the geometry is really a *hyper*graph.

**Reply by vishnu564** [reply]
How can we extract the indices, bonds/edges of each index, bond dimension of each edge from the representation?

**Reply by jcmgray** [reply]
You can just call `tn.ind_sizes()`, see [here](https://quimb.readthedocs.io/en/latest/autoapi/quimb/tensor/tensor_core/index.html#quimb.tensor.tensor_core.TensorNetwork.ind_sizes), if you want a dict of index -> size.

I'm not entirely clear on the question, but it might be worth noting that the bond names and sizes are explicitly the index names and array dimensions of the tensor objects, there is no separate bond object.

**Comment by zadnarock** [comment]
Thank you very much for the detailed answer. I want to explore other tensor network contraction algorithms thereby wanting to use the network generated in some other framework. Can any random quantum circuit can be converted into a tensor network using quimb, if yes can all of them be contracted to obtain the final measurement state?

**Reply by jcmgray** [reply]
* Most common gates are supported, and its easy to add new ones or use the gate in raw form, so yes most arbitrary circuits can be constructed. Things like dynamic measurement are not supported yet however.
* Not 100% sure what you mean by 'final measurement state' but there are lots of things you can do listed here - https://quimb.readthedocs.io/en/latest/tensor-circuit.html#forming-the-target-tensor-network, including contracting the 'dense', single vector representation of the final state (though this is not scalable with number of qubits)
